### DISCUSSION TITLE
K3s not using EvictHard in KubeletConfiguration

### DISCUSSION THREAD

**Discussion by JasperWey** [k3s-io/k3s#7581] [Q&A]
**Environmental Info:**
K3s Version: v1.27.1+k3s1

Node(s) CPU architecture, OS, and Version:  x86 cenots7

Cluster Configuration:

`ExecStart=/usr/local/bin/k3s \
    server \
        '--write-kubeconfig' \
        '/root/.kube/config' \
        '--cluster-cidr' \
        '113.122.0.0/16' \
        '--service-cidr' \
        '113.123.0.0/16' \
        '--service-node-port-range' \
        '400-32767' \
        '--node-name' \
        'uap-node' \
        '--docker' \
        '--disable' \
        'traefik' \
        '--kube-proxy-arg=proxy-mode=ipvs' \
        '--default-local-storage-path' \
        '/dataLog/persist/pvroot' \
        '--kubelet-arg=config=/etc/rancher/k3s/kubelet.config' \`

and the kubelet.config content:
`apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
imageGCHighThresholdPercent: 100
containerLogMaxSize: 100Mi
containerLogMaxFiles: 5
evictionHard:
  nodefs.available: "0%"
  nodefs.inodesFree: "0%"
  imagefs.available: "0%"
`

**Describe the bug:**

the nodefs.available setting in kubelet.conf not apply in kubelet. using default setting :

`May 18 19:45:48 node-131 k3s[17053]: time="2023-05-18T19:45:48+08:00" level=info msg="Running kubelet --address=0.0.0.0 --allowed-unsafe-sysctls=net.ipv4.ip_forward,net.ipv6.conf.all.forwarding --anonymous-auth=false --authentication-token-webhook=true --authorization-mode=Webhook --cgroup-driver=cgroupfs --client-ca-file=/var/lib/rancher/k3s/agent/client-ca.crt --cloud-provider=external --cluster-dns=113.123.0.10 --cluster-domain=cluster.local --config=/etc/rancher/k3s/kubelet.config --container-runtime-endpoint=unix:///run/k3s/cri-dockerd/cri-dockerd.sock **--eviction-hard=imagefs.available<5%,nodefs.available<5%** --eviction-minimum-reclaim=imagefs.available=10%,nodefs.available=10% --fail-swap-on=false --healthz-bind-address=127.0.0.1 --hostname-override=uap-node --kubeconfig=/var/lib/rancher/k3s/agent/kubelet.kubeconfig --node-labels= --pod-infra-container-image=rancher/mirrored-pause:3.6 --pod-manifest-path=/var/lib/rancher/k3s/agent/pod-manifests --read-only-port=0 --resolv-conf=/etc/resolv.conf --serialize-image-pulls=false --tls-cert-file=/var/lib/rancher/k3s/agent/serving-kubelet.crt --tls-private-key-file=/var/lib/rancher/k3s/agent/serving-kubelet.key"
`
**Steps To Reproduce:**

- Installed K3s: 
- make disk usage up to 98%

**Expected behavior:**

no image or container evict

**Actual behavior:**

images delete for clearing disk space

**Additional context / logs:**

**Comment by JasperWey** [comment]
when i using `'--kubelet-arg=eviction-hard=imagefs.available<1%,nodefs.available<1%' \`. it works. but i want to disable evict, so need to set to 0, but it not supproted

reference:
https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/#kubelet-config-k8s-io-v1beta1-KubeletConfiguration
`evictionHard is a map of signal names to quantities that defines hard eviction thresholds. For example: {"memory.available": "300Mi"}. To explicitly disable, pass a 0% or 100% threshold on an arbitrary resource. Default: memory.available: "100Mi" nodefs.available: "10%" nodefs.inodesFree: "5%" imagefs.available: "15%"`

**Reply by brandond** [reply]
`--kubelet-arg=config` is difficult to use, as the kubelet ignores all the other CLI flags if you specify a config file. You need to convert **all** the other CLI flags into your config file.

> but i want to disable evict, so need to set to 0, but it not supproted

I'm not sure what you mean it's not supported, the docs you just quoted show you that it is supported, and how to do it. Like this: `--kubelet-arg 'eviction-hard=imagefs.available<0%,memory.available<0%,nodefs.available<0%'`

**Reply by JasperWey** [reply]
K3s.service would faild as imagefs.avilable value should be positive, can not be set to <0%。setting **imagefs.avilable <0.000001%** and it works.

**Reply by brandond** [reply]
Huh, what I put above worked for me no problem. I literally ran it before pasting it into my comment.
